File: src/services/code_action/__tests__/insert_type_utils_tests.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open OUnit2
open Ty

let stylize = (new Insert_type_utils.stylize_ty_mapper ())#on_t Loc.none

let tests =
  "insert_type_utils"
  >::: [
         ( "stylize_union_number_with_number_literal" >:: fun ctxt ->
           let t_in = Union (false, Num, NumLit "1", []) in
           let t_exp = Num in
           assert_equal ~ctxt ~printer:Ty.show t_exp (stylize t_in)
         );
         ( "stylize_union_string_with_string_literal" >:: fun ctxt ->
           let t_in = Union (false, StrLit (Reason.OrdinaryName "foo"), Str, []) in
           let t_exp = Str in
           assert_equal ~ctxt ~printer:Ty.show t_exp (stylize t_in)
         );
         ( "stylize_union_true_and_false" >:: fun ctxt ->
           let t_in = Union (false, BoolLit true, BoolLit false, []) in
           let t_exp = Bool in
           assert_equal ~ctxt ~printer:Ty.show t_exp (stylize t_in)
         );
         ( "stylize_union_true_and_bool" >:: fun ctxt ->
           let t_in = Union (false, BoolLit true, Bool, []) in
           let t_exp = Bool in
           assert_equal ~ctxt ~printer:Ty.show t_exp (stylize t_in)
         );
         ( "stylize_union_string_number_literals" >:: fun ctxt ->
           let t_in = Union (false, Str, NumLit "1", [NumLit "2"]) in
           let t_exp = Union (false, NumLit "1", NumLit "2", [Str]) in
           assert_equal ~ctxt ~printer:Ty.show t_exp (stylize t_in)
         );
         (* These tests just document that sorting is working in a sane order *)
         ( "sort_types_numeric_literals" >:: fun ctxt ->
           let t_in = Union (false, NumLit "5", NumLit "11", [NumLit "1"; NumLit "2"]) in
           let t_exp = Union (false, NumLit "1", NumLit "2", [NumLit "5"; NumLit "11"]) in
           assert_equal ~ctxt ~printer:Ty.show t_exp (Insert_type.simplify t_in)
         );
         ( "sort_types_top_any" >:: fun ctxt ->
           let t_in = Union (false, Top, Any (Annotated ALoc.none), []) in
           let t_exp = Top in
           assert_equal ~ctxt ~printer:Ty.show t_exp (Insert_type.simplify t_in)
         );
         ( "sort_types_bot_any" >:: fun ctxt ->
           let t_in = Union (false, Bot EmptyType, Any (Annotated ALoc.none), []) in
           let t_exp = Any (Annotated ALoc.none) in
           assert_equal ~ctxt ~printer:Ty.show t_exp (Insert_type.simplify t_in)
         );
         ( "sort_types_any_first" >:: fun ctxt ->
           let t_in =
             Union (false, Void, Any (Annotated ALoc.none), [Null; Str; NumLit "5"; Bool])
           in
           let t_exp =
             Union (false, Any (Annotated ALoc.none), Void, [Null; Bool; NumLit "5"; Str])
           in
           assert_equal ~ctxt ~printer:Ty.show t_exp (Insert_type.simplify t_in)
         );
       ]

File: rust_port/crates/flow_services_code_action/src/insert_type_utils_tests.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! Port of `services/code_action/__tests__/insert_type_utils_tests.ml`

use std::sync::Arc;

use flow_aloc::ALoc;
use flow_common::reason::Name;
use flow_common_ty::ty::AnyKind;
use flow_common_ty::ty::BotKind;
use flow_common_ty::ty::Ty;
use flow_parser::loc::Loc;
use flow_parser::loc_sig::LocSig;

use crate::insert_type;
use crate::insert_type_utils::StylizeTyMapper;

fn stylize(t: Arc<Ty<ALoc>>) -> Arc<Ty<ALoc>> {
    use flow_common_ty::ty::TyEndoTy;
    let mut mapper = StylizeTyMapper::new();
    mapper.on_t(&Loc::default(), t)
}

//   >::: [

// ( "Stylize union: number with number literal" >:: fun ctxt ->
// );
#[test]
fn stylize_union_number_with_number_literal() {
    let t_in = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::Num),
        Arc::new(Ty::NumLit("1".into())),
        Arc::from([]),
    ));
    let t_exp = Arc::new(Ty::Num);
    assert_eq!(t_exp, stylize(t_in));
}

// ( "Stylize union: string with string literal" >:: fun ctxt ->
// );
#[test]
fn stylize_union_string_with_string_literal() {
    let t_in = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::StrLit(Name::new("foo"))),
        Arc::new(Ty::Str),
        Arc::from([]),
    ));
    let t_exp = Arc::new(Ty::Str);
    assert_eq!(t_exp, stylize(t_in));
}

// ( "Stylize union: true and false" >:: fun ctxt ->
// );
#[test]
fn stylize_union_true_and_false() {
    let t_in = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::BoolLit(true)),
        Arc::new(Ty::BoolLit(false)),
        Arc::from([]),
    ));
    let t_exp = Arc::new(Ty::Bool);
    assert_eq!(t_exp, stylize(t_in));
}

// ( "Stylize union: true and bool" >:: fun ctxt ->
// );
#[test]
fn stylize_union_true_and_bool() {
    let t_in = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::BoolLit(true)),
        Arc::new(Ty::Bool),
        Arc::from([]),
    ));
    let t_exp = Arc::new(Ty::Bool);
    assert_eq!(t_exp, stylize(t_in));
}

// ( "Stylize union: string and number literals" >:: fun ctxt ->
// );
#[test]
fn stylize_union_string_number_literals() {
    let t_in = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::Str),
        Arc::new(Ty::NumLit("1".into())),
        Arc::from([Arc::new(Ty::NumLit("2".into()))]),
    ));
    let t_exp = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::NumLit("1".into())),
        Arc::new(Ty::NumLit("2".into())),
        Arc::from([Arc::new(Ty::Str)]),
    ));
    assert_eq!(t_exp, stylize(t_in));
}

// ( "Sort types: numeric literals" >:: fun ctxt ->
// );
#[test]
fn sort_types_numeric_literals() {
    let t_in = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::NumLit("5".into())),
        Arc::new(Ty::NumLit("11".into())),
        Arc::from([
            Arc::new(Ty::NumLit("1".into())),
            Arc::new(Ty::NumLit("2".into())),
        ]),
    ));
    let t_exp = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::NumLit("1".into())),
        Arc::new(Ty::NumLit("2".into())),
        Arc::from([
            Arc::new(Ty::NumLit("5".into())),
            Arc::new(Ty::NumLit("11".into())),
        ]),
    ));
    assert_eq!(t_exp, insert_type::simplify(t_in));
}

// ( "Sort types: top, any" >:: fun ctxt ->
// );
#[test]
fn sort_types_top_any() {
    let t_in = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::Top),
        Arc::new(Ty::Any(AnyKind::Annotated(ALoc::none()))),
        Arc::from([]),
    ));
    let t_exp = Arc::new(Ty::Top);
    assert_eq!(t_exp, insert_type::simplify(t_in));
}

// ( "Sort types: bot, any" >:: fun ctxt ->
// );
#[test]
fn sort_types_bot_any() {
    let t_in = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::Bot(BotKind::EmptyType)),
        Arc::new(Ty::Any(AnyKind::Annotated(ALoc::none()))),
        Arc::from([]),
    ));
    let t_exp = Arc::new(Ty::Any(AnyKind::Annotated(ALoc::none())));
    assert_eq!(t_exp, insert_type::simplify(t_in));
}

// ( "Sort types: any first" >:: fun ctxt ->
//     Union (false, Void, Any (Annotated ALoc.none), [Null; Str; NumLit "5"; Bool])
//     Union (false, Any (Annotated ALoc.none), Void, [Null; Bool; NumLit "5"; Str])
// );
#[test]
fn sort_types_any_first() {
    let t_in = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::Void),
        Arc::new(Ty::Any(AnyKind::Annotated(ALoc::none()))),
        Arc::from([
            Arc::new(Ty::Null),
            Arc::new(Ty::Str),
            Arc::new(Ty::NumLit("5".into())),
            Arc::new(Ty::Bool),
        ]),
    ));
    let t_exp = Arc::new(Ty::Union(
        false,
        Arc::new(Ty::Any(AnyKind::Annotated(ALoc::none()))),
        Arc::new(Ty::Void),
        Arc::from([
            Arc::new(Ty::Null),
            Arc::new(Ty::Bool),
            Arc::new(Ty::NumLit("5".into())),
            Arc::new(Ty::Str),
        ]),
    ));
    assert_eq!(t_exp, insert_type::simplify(t_in));
}
